src/js/components/Carousel/Carousel.js

```diff
@@ -1,242 +1,222 @@
-  onFocus,
-  onBlur,
-  ...rest
+  width,
+  initialChild,
+  onChild,
+  play,
 }) => {
-  const theme = useContext(ThemeContext) || defaultProps.theme;
-  const [focus, setFocus] = useState();
-  const timerRef = useRef();
-
-  const [indexes, setIndexes] = useState({
-    activeIndex: activeChild !== undefined ? activeChild : initialChild,
+  const noContainer = !fill && !height && !width;
+  const numSlides = children.length;
+  const firstChildRef = useRef(null);
+  const [current, setCurrent] = useState(initialChild);
+  const [previous, setPrevious] = useState(undefined);
+  const [containerProps, setContainerProps] = useState({
+    heightProp: height,
+    widthProp: width,
+  });
+  const [direction, setDirection] = useState(undefined);
+  const [inTransition, setInTransition] = useState(false);
+  const onJumpNavigation = handleOnJumpNavigation({
+    setDirection,
+    setCurrent,
+    setPrevious,
+    setInTransition,
+    onChild,
+  });
+  const onNext = handleOnNext({
+    setDirection,
+    setCurrent,
+    setPrevious,
+    setInTransition,
+    numSlides,
+    onChild,
+  });
+  const onPrevious = handleOnPrevious({
+    setDirection,
+    setCurrent,
+    setPrevious,
+    setInTransition,
+    numSlides,
+    onChild,
   });
 
-  const { activeIndex, priorActiveIndex } = indexes;
-  const lastIndex = Children.count(children) - 1;
-
-  if (activeIndex !== activeChild && activeChild !== undefined) {
-    if (activeChild >= 0 && activeChild <= lastIndex) {
-      setIndexes({
-        activeIndex: activeChild,
-        priorActiveIndex: activeIndex,
+  /**
+   * Handles when the "fill" or "height" & "width" props are not specified
+   * and no Carousel container is allocated for the slides. This renders
+   * the first active child to expand to the given height and width it
+   * needs and then dynamically reads those values to set as the fixed
+   * dimensions of the Carousel container.
+   */
+  useEffect(() => {
+    if (
+      noContainer &&
+      (containerProps.heightProp === undefined ||
+        containerProps.widthProp === undefined)
+    ) {
+      const { current: childRef } = firstChildRef;
+      if (childRef) {
+        if (childRef.offsetWidth > 0 && childRef.offsetHeight > 0) {
+          setContainerProps({
+            heightProp: `${childRef.offsetHeight}px`,
+            widthProp: `${childRef.offsetWidth}px`,
+          });
+        } else setContainerProps({});
+      }
+    } else {
+      setContainerProps({
+        heightProp: fill ? '100%' : height,
+        widthProp: fill ? '100%' : width,
       });
     }
-  }
+  }, [noContainer, fill, height, width]);
 
-  const onChildChange = useCallback(
-    (index) => {
-      if (onChild) {
-        onChild(index);
-      }
-    },
-    [onChild],
-  );
+  /**
+   * Delays the transitions between Carousel slides. This is needed to
+   * avoid users "spamming" the controls which results in jarring animations
+   * and a bad user experience.
+   */
```

Comment:
Thank you for including comments accompanying useEffects throughout. Very helpful.